### PR TITLE
cl_ext_cxx_for_opencl

### PR DESCRIPTION
**Pull request by kpet**
Draft specification to expose C++ for OpenCL as a supported language.

This change also allocates an enum block for EXT extensions.

Signed-off-by: Kevin Petit <kevin.petit@arm.com>

### CODE REVIEW

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
++
+--
+
+[cols="1,1,4",options="header"]
+|====
+| cl_device_info
+| Return Type
+| Description
+
+| `CL_DEVICE_CXX_FOR_OPENCL_VERSION_EXT`
```
[the remark this answers is not in the document]
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:91]: Yes, this was a deliberate choice. Since the language is still young/experimental and no versions have been formally ratified, we thought reporting a single version made more sense. We don't want to set expectations with respect to support for older versions.

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+| `cl_version`
+| Returns the version of the C++ for OpenCL specification implemented by the
+  device compiler.
```
[the remark this answers is not in the document]
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:94]: We touched on that during a call. We'd want to tag OpenCL-Docs with versions of C++ for OpenCL. It would have its own independent version number. We don't think mandating anything makes sense at this stage given the current "moving target" nature of the language. As to what the current version is, the spec currently says that it would be https://github.com/KhronosGroup/OpenCL-Docs/blob/master/cxx4opencl/diff2openclc.txt#L147 but we may want to use a smaller number.

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+== Overview
+
+This extension adds support for building programs written using the C++ for
+OpenCL kernel language. It also enables applications to query the version of
+the language supported by the device compiler.
+
+== New build option
+
+This extension adds support for a new `CLC++` value to be passed to the
```
[the remark this answers is not in the document]
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:59]: We feel it's a bit early to have a version number in the build option but that's certainly something we'd want to do in the future as the language stabilises (potentially as part of a ratification).

[the remark this answers is not in the document]
Comment (alycm) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:59]: Yes, okay, `-cl-std=CLC++` seems fine for now.  I'll resolve this comment.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (alycm) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:94]: The current C++ for OpenCL document says "version v3.0.3-rc" at the top, so without something to override that, like the OpenCL C spec does, a reader is likely to assume that is the version number.

But for the general principal of no number while a "moving target", okay.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (alycm) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:91]: Okay.

**Review by alycm — APPROVED** [review]

File: xml/cl.xml [hunk 4]
```diff
@@ -1787,8 +1787,13 @@ server's OpenCL/api-docs repository.
             <unused start="0x4214" end="0x421F"/>
     </enums>
 
-    <enums start="0x4220" end="0x4FFF" name="enums.4220" comment="Reserved for vendor extensions. Allocate in groups of 16.">
-            <unused start="0x420F" end="0x4FFF"/>
+    <enums start="0x4220" end="0x422F" name="enums.4220" comment="Reserved for EXT extensions">
+        <enum value="0x4220"        name="CL_DEVICE_CXX_FOR_OPENCL_VERSION_EXT"/>
```
Comment (alycm) [inline xml/cl.xml:1791]: I've just realized that I merged an XML change that reserves this block: https://github.com/KhronosGroup/OpenCL-Docs/pull/411

Hopefully it's not an issue to change this number?  Sorry I didn't make the connection despite reviewing both PRs.  If there are concerns about it happening again before this branch merges then the XML change could be its own PR.

Comment (kpet) [reply xml/cl.xml]: I've changed the allocation. This should be fine.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:94]: Yes, that's a good point. We probably ought to change this. Thanks for flagging.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnastasiaStulova) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:91]: However, if we use `CL_DEVICE_CXX_FOR_OPENCL_ALL_VERSIONS_EXT` we don't lose anything and in case we start to version the language we won't need to propose yet another extension or a version of an extension which is obviously more complicated for vendors and developers. It will also make things better aligned with the general direction.

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+[cols="1,1,4",options="header"]
+|====
+| cl_device_info
+| Return Type
+| Description
+
+| `CL_DEVICE_CXX_FOR_OPENCL_VERSION_EXT`
+| `cl_version`
+| Returns the version of the C++ for OpenCL specification implemented by the
```
Comment (AnastasiaStulova) [inline extensions/cl_ext_cxx_for_opencl.asciidoc:93]:
```suggestion
| Returns the version of the C++ for OpenCL language supported by the
```

Comment (AnastasiaStulova) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:93]: I think it's better to change this since we don't have the specification formally.

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+This extension is written against the OpenCL Specification
+Version 3.0.2.
+
+This extension requires OpenCL 3.0 or OpenCL 2.x and `cl_khr_extended_versioning`.
+
+== Overview
+
+This extension adds support for building programs written using the C++ for
+OpenCL kernel language. It also enables applications to query the version of
+the language supported by the device compiler.
```
Comment (AnastasiaStulova) [inline extensions/cl_ext_cxx_for_opencl.asciidoc:55]: Since the language documentation is not published in the register do we want to add a link to pdf blob:
https://github.com/KhronosGroup/Khronosdotorg/blob/master/api/opencl/assets/CXX_for_OpenCL.pdf

Although maybe it's also time to think of a better way to publish this and a process.

FYI, we will probably attach pdf as an artifact once we create a tag for C++ for OpenCL on GitHub?

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:91]: Discussed with Anastasia offline: we decided to report a single version as per my message above but rename the query to `CL_DEVICE_CXX_FOR_OPENCL_NUMERIC_VERSION_EXT` for consistency with `cl_khr_extended_versioning`.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 5 above)
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:93]: Yes, I think I prefer that as well.

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+include::../copyrights.txt[]
+
+== Status
+
+Draft
+
+== Version
+
+Built On: {docdate} +
+Version: 0.1.0
```
Comment (AnastasiaStulova) [inline extensions/cl_ext_cxx_for_opencl.asciidoc:42]: Do we need to update version here too?

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+Marco Antognini, Arm Ltd. +
+Neil Hickey, Arm Ltd. +
+
+== Notice
+
+include::../copyrights.txt[]
+
+== Status
+
+Draft
```
Comment (AnastasiaStulova) [inline extensions/cl_ext_cxx_for_opencl.asciidoc:37]: When will we change the status?

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 7 above)
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:42]: Yup, good catch, it would be nice if we could make that automatic.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 8 above)
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:37]: The status makes less sense for an EXT (no formal approval or ratification), I'll remove the section entirely.

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+Built On: {docdate} +
+Version: 1.0.0
+
+== Dependencies
+
+This extension is written against the OpenCL Specification
+Version 3.0.2.
+
+This extension requires OpenCL 3.0 or OpenCL 2.x and `cl_khr_extended_versioning`.
```
Comment (alycm) [inline extensions/cl_ext_cxx_for_opencl.asciidoc:45]: > This extension requires OpenCL 3.0 or **OpenCL 2.x** and `cl_khr_extended_versioning`.

Emphasis mine.  Mostly out of curiosity rather than because I have a need for it, but is there any reason why OpenCL 1.2 and `cl_khr_extended_versioning` wouldn't be possible?

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+OpenCL kernel language. It also enables applications to query the version of
+the language supported by the device compiler.
+
+== New build option
+
+This extension adds support for a new `CLC++` value to be passed to the
+`-cl-std` build option accepted by *clBuildProgram* and *clCompileProgram*.
+
+This option is only valid for programs created using
+*clCreateProgramFromSource*.
```
Comment (alycm) [inline extensions/cl_ext_cxx_for_opencl.asciidoc:59]: The definition of `clCreateProgramFromSource` says:

> The devices associated with the program object are the devices associated with context. The source code specified by strings is either an OpenCL C program source, header or implementation-defined source for custom devices that support an online compiler. OpenCL C++ is not supported as an online-compiled kernel language through this interface.

Should this extension say something about causing `clCreateProgramFromSource` to accept C++ for OpenCL source?

**Review by alycm — APPROVED** [review]
LGTM again.  One naming correction, and an additional though from the reread.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (alycm) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:90]: Given the problem actually exists outside of this extension point I think this comment can be resolved by just filing an issue,

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (alycm) [reply extensions/cl_ext_cxx_for_opencl.asciidoc]: Okay, makes sense to me.  I'll resolve this.

File: xml/cl.xml [hunk 11]
```diff
@@ -5651,5 +5656,10 @@ server's OpenCL/api-docs repository.
                 <enum name="CL_DEVICE_NODE_MASK_KHR"/>
             </require>
         </extension>
+        <extension name="cl_ext_cxx_for_opencl" supported="opencl">
+            <require comment="cl_device_info">
+                <enum name="CL_DEVICE_CXX_FOR_OPENCL_VERSION_EXT"/>
```
[the remark this answers is not in the document]
Comment (kpet) [reply xml/cl.xml:5661]: Yes, it's a shame the tools don't help here. I've created https://github.com/KhronosGroup/OpenCL-Docs/issues/419

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 9 above)
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:45]: C++ for OpenCL depends on OpenCL C 2.0 at the moment. We haven't really looked at how it interacts with OpenCL C 3.0 and optional language features. I'll add an explicit requirement on OpenCL C 2.0 here.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 10 above)
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:59]: This is implied by the current text but being more explicit wouldn't hurt. I'll add something.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc:90]: I've created https://github.com/KhronosGroup/OpenCL-Docs/issues/420 to cover what we discussed in the last teleconference.

File: extensions/cl_ext_cxx_for_opencl.asciidoc (hunk 10 above)
Comment (alycm) [reply extensions/cl_ext_cxx_for_opencl.asciidoc]: Yeah, I definitely agree that it's implied.

File: extensions/cl_ext_cxx_for_opencl.asciidoc [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+Built On: {docdate} +
+Version: 0.1.0
+
+== Dependencies
+
+This extension is written against the OpenCL Specification
+Version 3.0.2.
+
+This extension requires OpenCL 3.0 or OpenCL 2.x and `cl_khr_extended_versioning`.
```
[the remark this answers is not in the document]
Comment (AnastasiaStulova) [reply extensions/cl_ext_cxx_for_opencl.asciidoc]: Are we actually using anything from `cl_khr_extended_versioning` for this extension? Maybe it's better to say that functionality here is related to `cl_khr_extended_versioning` or expected to follow similar principles?

If it's possible to avoid an extra dependency it would be better.

[the remark this answers is not in the document]
Comment (alycm) [reply extensions/cl_ext_cxx_for_opencl.asciidoc]: `cl_version` comes from `cl_khr_extended_versioning` (well, technically that is `cl_version_khr`, but same thing).

[the remark this answers is not in the document]
Comment (kpet) [reply extensions/cl_ext_cxx_for_opencl.asciidoc]: The `cl_version_khr` type was introduced by `cl_khr_extended_versioning`. That's the reason for the dependency.
